msm/src/ffa/interpreter.rs

```diff
@@ -108,3 +119,134 @@ pub fn test_addition<F: PrimeField, Env: FFAInterpreterEnv<F>>(env: &mut Env, a:
+        a + b - q * f - r - c0 * limb_size
+    } else if limb_num < N_LIMBS - 1 {
+        let limb_size = Env::constant(From::from((1 << LIMB_BITSIZE) as u64));
+        let c_prev: Env::Variable = Env::read_column(env, FFAColumnIndexer::Carry(limb_num - 1));
+        let c_cur: Env::Variable = Env::read_column(env, FFAColumnIndexer::Carry(limb_num));
+        env.range_check_abs1(&c_prev);
+        env.range_check_abs1(&c_cur);
+        a + b - q * f - r - c_cur * limb_size + c_prev
+    } else {
+        let c_prev: Env::Variable = Env::read_column(env, FFAColumnIndexer::Carry(limb_num - 1));
+        env.range_check_abs1(&c_prev);
+        a + b - q * f - r + c_prev
+    };
+    env.assert_zero(constraint);
+}
+
+pub fn constrain_ff_addition<F: PrimeField, Env: FFAInterpreterEnv<F>>(env: &mut Env) {
+    for limb_i in 0..N_LIMBS {
+        constrain_ff_addition_row(env, limb_i);
+    }
+}
+
+pub fn ff_addition_circuit<F: PrimeField, Ff: PrimeField, Env: FFAInterpreterEnv<F>>(
+    env: &mut Env,
+    a: Ff,
+    b: Ff,
+) {
+    let f_bigint: BigUint = TryFrom::try_from(Ff::Params::MODULUS).unwrap();
+
+    let a_limbs: [F; N_LIMBS] = limb_decompose_ff(&a);
+    let b_limbs: [F; N_LIMBS] = limb_decompose_ff(&b);
+    let f_limbs: [F; N_LIMBS] = limb_decompose_bui(f_bigint.clone());
+    a_limbs.iter().enumerate().for_each(|(i, var)| {
+        env.copy(
+            &Env::constant(*var),
+            Env::column_pos(FFAColumnIndexer::InputA(i)),
+        );
+    });
+    b_limbs.iter().enumerate().for_each(|(i, var)| {
+        env.copy(
+            &Env::constant(*var),
+            Env::column_pos(FFAColumnIndexer::InputB(i)),
+        );
+    });
+    f_limbs.iter().enumerate().for_each(|(i, var)| {
+        env.copy(
+            &Env::constant(*var),
+            Env::column_pos(FFAColumnIndexer::ModulusF(i)),
+        );
+    });
+
+    let a_bigint = FieldHelpers::to_biguint(&a);
+    let b_bigint = FieldHelpers::to_biguint(&b);
+
+    // TODO FIXME this computation must be done over BigInts, not BigUInts
+    // q can be -1! But only in subtraction, so for now we don't care.
+    // for now with addition only q ∈ {0,1}
+    let (q_bigint, r_bigint) = (a_bigint + b_bigint).div_rem(&f_bigint);
+    let r_limbs: [F; N_LIMBS] = limb_decompose_bui(r_bigint);
+    // We expect just one limb.
+    let q: F = limb_decompose_bui(q_bigint)[0];
+
+    env.copy(
+        &Env::constant(q),
+        Env::column_pos(FFAColumnIndexer::Quotient),
+    );
+    r_limbs.iter().enumerate().for_each(|(i, var)| {
+        env.copy(
+            &Env::constant(*var),
+            Env::column_pos(FFAColumnIndexer::Remainder(i)),
+        );
+    });
+
+    let limb_size: F = From::from((1 << LIMB_BITSIZE) as u64);
+    let mut carry: F = From::from(0u64);
+    for limb_i in 0..N_LIMBS {
+        let res = a_limbs[limb_i] + b_limbs[limb_i] - q * f_limbs[limb_i] - r_limbs[limb_i] + carry;
+        let newcarry: F = if res == limb_size {
```

Comment:
Don't you want to compute and constraint carry in the circuit too?

Reply:
I do! I even left a comment above :)
```
// For now this function does not /compute/ anything, although it could.
```

But maybe in a follow-up PR?